packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,568 @@
+
+  List<Paint> get fillPaint => _fillPaint;
+  final List<Paint> _fillPaint = <Paint>[];
+
+  static Size _boxSize(RenderBox box) => box == null ? Size.zero : box.size;
+
+  static Rect _boxRect(RenderBox box) {
+    final _SegmentedControlContainerBoxParentData childData = box.parentData;
+    return box == null ? Rect.zero : childData.offset & box.size;
+  }
+
+  static double _minWidth(RenderBox box, double height) {
+    return box == null ? 0.0 : box.getMinIntrinsicWidth(height);
+  }
+
+  static double _minHeight(RenderBox box, double width) {
+    return box == null ? 0.0 : box.getMinIntrinsicHeight(width);
+  }
+
+  @override
+  double computeMinIntrinsicWidth(double height) {
+    double total = 0.0;
+    RenderBox child = firstChild;
+    while (child != null) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      total += _minWidth(child, height);
+      child = childParentData.nextSibling;
+    }
+    return total;
+  }
+
+  @override
+  double computeMaxIntrinsicWidth(double height) => computeMinIntrinsicWidth(height);
+
+  @override
+  double computeMinIntrinsicHeight(double width) => _minHeight(firstChild, width);
+
+  @override
+  double computeMaxIntrinsicHeight(double width) => computeMinIntrinsicHeight(width);
+
+  @override
+  double computeDistanceToActualBaseline(TextBaseline baseline) {
+    return firstChild.getDistanceToActualBaseline(baseline);
+  }
+
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! _SegmentedControlContainerBoxParentData) {
+      child.parentData = new _SegmentedControlContainerBoxParentData();
+    }
+  }
+
+  @override
+  void performLayout() {
+    double maxHeight = _kMinSegmentedControlHeight;
+    final Constraints parentConstraints = constraints.loosen();
+
+    int length = 0;
+    RenderBox child = firstChild;
+
+    while (child != null) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      child.layout(parentConstraints, parentUsesSize: true);
+      final double boxHeight = child.size.height;
+      if (boxHeight > maxHeight) {
+        maxHeight = boxHeight;
+      }
+      child = childParentData.nextSibling;
+      length += 1;
+    }
+
+    final double overallWidth = constraints.maxWidth;
+    final Size overallSize = constraints.constrain(Size(overallWidth, maxHeight));
+
+    Offset centerLayout(Size boxSize, double x) {
+      assert(maxHeight >= boxSize.height);
+      final Offset boxOffset = new Offset(x, (maxHeight - boxSize.height) / 2.0);
+      return boxOffset;
+    }
+
+    final double childWidth = overallSize.width / length;
+
+    final BoxConstraints childConstraints =
+      new BoxConstraints.tightFor(width: childWidth, height: maxHeight);
+
+    child = firstChild;
+    while (child != null) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      child.layout(childConstraints, parentUsesSize: true);
```

Comment:
you've already laid out the child once. laying out the child again makes this an O(N^2) algorithm instead of O(N). We generally want to work really really hard to avoid this.

If you really need to first check the child's dimensions then lay them out at different dimensions, use the intrinsic size protocol. But in general I would discourage this. Either pick a size and force the children to that size (e.g. `constraints.constraintHeight(_kMinSegmentedControlHeight)`), or let the children size their own height arbitrarily, then align them so they're all centered.

Reply:
(doing layout twice with different constraints guarantees that you will lose all the performance optimisations that short-circuit child layout when the constraints haven't changed)